0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+| `expiresAt` | Fixed-Length [Interledger Timestamp](../asn1/InterledgerTypes.asn) | Date and time when the packet expires. This field is modified by each connector, who subtracts some amount of time from the expiry before forwarding the packet |
+| `executionCondition` | UInt256 | Sha256 hash digest of the `fulfillment` that will execute the transfer of value represented by this packet |
+| `destination` | [ILP Address](../0015-ilp-addresses/0015-ilp-addresses.md) | ILP Address of the receiver |
+| `data` | Variable-Length Octet String | End-to-end data. Connectors MUST NOT modify this data |
+
+See the ASN.1 definition [here](../asn1/InterledgerProtocol.asn).
+
+#### ILP Fulfill
+
+ILP Fulfill packets are `type` 13.
+
+| Field | Type | Description |
+|---|---|---|
+| `fulfillment` | UInt256 | 32-byte preimage of the `executionCondition` from the corresponding ILP Prepare packet | 
+| `data` | Variable-Length Octet String | End-to-end data. Connectors MUST NOT modify this data |
+
+See the ASN.1 definition [here](../asn1/InterledgerProtocol.asn).
+
+#### ILP Reject
+
+ILP Reject packets are `type` 14.
+
+| Field | Type | Description |
+|---|---|---|
+| `code` | 3-Character IA5String | [ILP Error Code](#error-codes) |
+| `triggeredBy` | [ILP Address](../0015-ilp-addresses/0015-ilp-addresses.md) | ILP Address of the party that created this error |
+| `message` | UTF8 String | User-readable error message, primarily intended for debugging purposes |
+| `data` | Variable-Length Octet String | End-to-end data. Connectors MUST NOT modify this data |
+
+See the ASN.1 definition [here](../asn1/InterledgerProtocol.asn).
+
+### Error Codes
+
+#### F__ - Final Error
+
+Final errors indicate that the payment is invalid and should not be retried unless the details are changed.
+
+| Code | Name | Description | Data Fields |
+|---|---|---|---|
+| **F00** | **Bad Request** | Generic sender error. | (empty) |
+| **F01** | **Invalid Packet** | The ILP packet was syntactically invalid. | (empty) |
+| **F02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or the connector does not have a route to the destination. | (empty) |
+| **F03** | **Invalid Amount** | The amount is invalid, for example it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. | (empty) |
+| **F04** | **Insufficient Destination Amount** | The receiver deemed the amount insufficient, for example you tried to pay a $100 invoice with $10. | (empty) |
+| **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. | (empty) |
+| **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) | (empty) |
```

Comment:
The description refers to a "memo" which is (I think) not defined elsewhere in the protocol.